File: packages/sdk-php/tests/ModelsTest.php
<?php

declare(strict_types=1);

namespace PocketPing\Tests;

use PHPUnit\Framework\TestCase;
use PocketPing\Models\ConnectRequest;
use PocketPing\Models\ConnectResponse;
use PocketPing\Models\CustomEvent;
use PocketPing\Models\IdentifyRequest;
use PocketPing\Models\Message;
use PocketPing\Models\MessageStatus;
use PocketPing\Models\ReadRequest;
use PocketPing\Models\SendMessageRequest;
use PocketPing\Models\Sender;
use PocketPing\Models\Session;
use PocketPing\Models\SessionMetadata;
use PocketPing\Models\TrackedElement;
use PocketPing\Models\TriggerOptions;
use PocketPing\Models\UserIdentity;
use PocketPing\Models\VersionCheckResult;
use PocketPing\Models\VersionStatus;
use PocketPing\Models\WebSocketEvent;

class ModelsTest extends TestCase
{
    // ─────────────────────────────────────────────────────────────────
    // UserIdentity Tests
    // ─────────────────────────────────────────────────────────────────

    public function testUserIdentityFromArray(): void
    {
        $identity = UserIdentity::fromArray([
            'id' => 'user-123',
            'email' => 'user@example.com',
            'name' => 'John Doe',
            'plan' => 'premium',
            'company' => 'Acme Inc',
        ]);

        $this->assertEquals('user-123', $identity->id);
        $this->assertEquals('user@example.com', $identity->email);
        $this->assertEquals('John Doe', $identity->name);
        $this->assertEquals('premium', $identity->customFields['plan']);
        $this->assertEquals('Acme Inc', $identity->customFields['company']);
    }

    public function testUserIdentityToArray(): void
    {
        $identity = new UserIdentity(
            id: 'user-123',
            email: 'user@example.com',
            name: 'John Doe',
            customFields: ['plan' => 'premium'],
        );

        $array = $identity->toArray();

        $this->assertEquals('user-123', $array['id']);
        $this->assertEquals('user@example.com', $array['email']);
        $this->assertEquals('John Doe', $array['name']);
        $this->assertEquals('premium', $array['plan']);
    }

    public function testUserIdentityRequiresId(): void
    {
        $this->expectException(\InvalidArgumentException::class);
        UserIdentity::fromArray(['email' => 'test@test.com']);
    }

    // ─────────────────────────────────────────────────────────────────
    // SessionMetadata Tests
    // ─────────────────────────────────────────────────────────────────

    public function testSessionMetadataFromArray(): void
    {
        $metadata = SessionMetadata::fromArray([
            'url' => 'https://example.com',
            'referrer' => 'https://google.com',
            'pageTitle' => 'Home Page',
            'userAgent' => 'Mozilla/5.0',
            'deviceType' => 'desktop',
        ]);

        $this->assertEquals('https://example.com', $metadata->url);
        $this->assertEquals('https://google.com', $metadata->referrer);
        $this->assertEquals('Home Page', $metadata->pageTitle);
        $this->assertEquals('desktop', $metadata->deviceType);
    }

    public function testSessionMetadataMergePreservesGeoInfo(): void
    {
        $original = new SessionMetadata(
            url: 'https://example.com/old',
            ip: '192.168.1.1',
            country: 'US',
            city: 'New York',
        );

        $new = new SessionMetadata(
            url: 'https://example.com/new',
            userAgent: 'Mozilla/5.0',
        );

        $merged = $original->mergeWith($new);

        $this->assertEquals('https://example.com/new', $merged->url);
        $this->assertEquals('192.168.1.1', $merged->ip);
        $this->assertEquals('US', $merged->country);
        $this->assertEquals('New York', $merged->city);
        $this->assertEquals('Mozilla/5.0', $merged->userAgent);
    }

    // ─────────────────────────────────────────────────────────────────
    // Session Tests
    // ─────────────────────────────────────────────────────────────────

    public function testSessionFromArray(): void
    {
        $session = Session::fromArray([
            'id' => 'session-123',
            'visitorId' => 'visitor-456',
            'createdAt' => '2024-01-01T12:00:00Z',
            'lastActivity' => '2024-01-01T13:00:00Z',
            'operatorOnline' => true,
            'aiActive' => false,
        ]);

        $this->assertEquals('session-123', $session->id);
        $this->assertEquals('visitor-456', $session->visitorId);
        $this->assertTrue($session->operatorOnline);
    }

    public function testSessionToArray(): void
    {
        $session = new Session(
            id: 'session-123',
            visitorId: 'visitor-456',
            createdAt: new \DateTimeImmutable('2024-01-01T12:00:00Z'),
            lastActivity: new \DateTimeImmutable('2024-01-01T13:00:00Z'),
            operatorOnline: true,
        );

        $array = $session->toArray();

        $this->assertEquals('session-123', $array['id']);
        $this->assertEquals('visitor-456', $array['visitorId']);
        $this->assertTrue($array['operatorOnline']);
    }

    // ─────────────────────────────────────────────────────────────────
    // Message Tests
    // ─────────────────────────────────────────────────────────────────

    public function testMessageFromArray(): void
    {
        $message = Message::fromArray([
            'id' => 'msg-123',
            'sessionId' => 'session-456',
            'content' => 'Hello!',
            'sender' => 'visitor',
            'timestamp' => '2024-01-01T12:00:00Z',
            'status' => 'sent',
        ]);

        $this->assertEquals('msg-123', $message->id);
        $this->assertEquals('session-456', $message->sessionId);
        $this->assertEquals('Hello!', $message->content);
        $this->assertEquals(Sender::VISITOR, $message->sender);
        $this->assertEquals(MessageStatus::SENT, $message->status);
    }

    public function testMessageWithStatus(): void
    {
        $message = new Message(
            id: 'msg-123',
            sessionId: 'session-456',
            content: 'Hello!',
            sender: Sender::VISITOR,
            timestamp: new \DateTimeImmutable(),
        );

        $updated = $message->withStatus(MessageStatus::READ);

        $this->assertEquals(MessageStatus::SENT, $message->status);
        $this->assertEquals(MessageStatus::READ, $updated->status);
    }

    // ─────────────────────────────────────────────────────────────────
    // TrackedElement Tests
    // ─────────────────────────────────────────────────────────────────

    public function testTrackedElementFromArray(): void
    {
        $element = TrackedElement::fromArray([
            'selector' => '#pricing-btn',
            'name' => 'clicked_pricing',
            'event' => 'click',
            'widgetMessage' => 'Need help with pricing?',
        ]);

        $this->assertEquals('#pricing-btn', $element->selector);
        $this->assertEquals('clicked_pricing', $element->name);
        $this->assertEquals('click', $element->event);
        $this->assertEquals('Need help with pricing?', $element->widgetMessage);
    }

    public function testTrackedElementDefaultEvent(): void
    {
        $element = TrackedElement::fromArray([
            'selector' => '#btn',
            'name' => 'test_event',
        ]);

        $this->assertEquals('click', $element->event);
    }

    // ─────────────────────────────────────────────────────────────────
    // CustomEvent Tests
    // ─────────────────────────────────────────────────────────────────

    public function testCustomEventFromArray(): void
    {
        $event = CustomEvent::fromArray([
            'name' => 'page_view',
            'data' => ['page' => '/pricing'],
            'timestamp' => '2024-01-01T12:00:00Z',
            'sessionId' => 'session-123',
        ]);

        $this->assertEquals('page_view', $event->name);
        $this->assertEquals('/pricing', $event->data['page']);
        $this->assertEquals('session-123', $event->sessionId);
    }

    public function testCustomEventWithSessionId(): void
    {
        $event = new CustomEvent(name: 'test_event');
        $updated = $event->withSessionId('session-123');

        $this->assertNull($event->sessionId);
        $this->assertEquals('session-123', $updated->sessionId);
    }

    // ─────────────────────────────────────────────────────────────────
    // ConnectRequest Tests
    // ─────────────────────────────────────────────────────────────────

    public function testConnectRequestFromArray(): void
    {
        $request = ConnectRequest::fromArray([
            'visitorId' => 'visitor-123',
            'sessionId' => 'session-456',
            'metadata' => ['url' => 'https://example.com'],
        ]);

        $this->assertEquals('visitor-123', $request->visitorId);
        $this->assertEquals('session-456', $request->sessionId);
        $this->assertEquals('https://example.com', $request->metadata->url);
    }

    // ─────────────────────────────────────────────────────────────────
    // SendMessageRequest Tests
    // ─────────────────────────────────────────────────────────────────

    public function testSendMessageRequestFromArray(): void
    {
        $request = SendMessageRequest::fromArray([
            'sessionId' => 'session-123',
            'content' => 'Hello!',
            'sender' => 'visitor',
        ]);

        $this->assertEquals('session-123', $request->sessionId);
        $this->assertEquals('Hello!', $request->content);
        $this->assertEquals(Sender::VISITOR, $request->sender);
    }

    public function testSendMessageRequestContentLengthLimit(): void
    {
        $longContent = str_repeat('a', 4001);

        $this->expectException(\InvalidArgumentException::class);
        $this->expectExceptionMessage('exceeds maximum length');

        new SendMessageRequest(
            sessionId: 'session-123',
            content: $longContent,
            sender: Sender::VISITOR,
        );
    }

    // ─────────────────────────────────────────────────────────────────
    // VersionCheckResult Tests
    // ─────────────────────────────────────────────────────────────────

    public function testVersionCheckResultToArray(): void
    {
        $result = new VersionCheckResult(
            status: VersionStatus::OUTDATED,
            message: 'Update available',
            minVersion: '0.2.0',
            latestVersion: '0.3.0',
            canContinue: true,
        );

        $array = $result->toArray();

        $this->assertEquals('outdated', $array['status']);
        $this->assertEquals('Update available', $array['message']);
        $this->assertEquals('0.2.0', $array['minVersion']);
        $this->assertEquals('0.3.0', $array['latestVersion']);
        $this->assertTrue($array['canContinue']);
    }

    // ─────────────────────────────────────────────────────────────────
    // WebSocketEvent Tests
    // ─────────────────────────────────────────────────────────────────

    public function testWebSocketEventToJson(): void
    {
        $event = new WebSocketEvent('message', ['content' => 'Hello']);

        $json = $event->toJson();
        $decoded = json_decode($json, true);

        $this->assertEquals('message', $decoded['type']);
        $this->assertEquals('Hello', $decoded['data']['content']);
    }

    // ─────────────────────────────────────────────────────────────────
    // JSON Serialization Tests
    // ─────────────────────────────────────────────────────────────────

    public function testModelsJsonSerializable(): void
    {
        $identity = new UserIdentity(id: 'user-123', email: 'test@test.com');
        $this->assertJson(json_encode($identity));

        $metadata = new SessionMetadata(url: 'https://example.com');
        $this->assertJson(json_encode($metadata));

        $session = new Session(
            id: 'session-123',
            visitorId: 'visitor-456',
            createdAt: new \DateTimeImmutable(),
            lastActivity: new \DateTimeImmutable(),
        );
        $this->assertJson(json_encode($session));

        $message = new Message(
            id: 'msg-123',
            sessionId: 'session-456',
            content: 'Hello!',
            sender: Sender::VISITOR,
            timestamp: new \DateTimeImmutable(),
        );
        $this->assertJson(json_encode($message));
    }
}

File: packages/sdk-go/models_test.go
package pocketping

import (
	"encoding/json"
	"testing"
	"time"
)

func TestUserIdentityMarshalJSON(t *testing.T) {
	identity := UserIdentity{
		ID:    "user-123",
		Email: "test@example.com",
		Name:  "Test User",
		Extra: map[string]interface{}{
			"plan":    "premium",
			"company": "Acme Inc",
		},
	}

	data, err := json.Marshal(identity)
	if err != nil {
		t.Fatalf("failed to marshal: %v", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}

	if result["id"] != "user-123" {
		t.Errorf("expected id=user-123, got %v", result["id"])
	}
	if result["email"] != "test@example.com" {
		t.Errorf("expected email=test@example.com, got %v", result["email"])
	}
	if result["plan"] != "premium" {
		t.Errorf("expected plan=premium, got %v", result["plan"])
	}
}

func TestUserIdentityUnmarshalJSON(t *testing.T) {
	data := []byte(`{"id":"user-456","email":"test@example.com","plan":"basic","company":"Test Corp"}`)

	var identity UserIdentity
	if err := json.Unmarshal(data, &identity); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}

	if identity.ID != "user-456" {
		t.Errorf("expected id=user-456, got %v", identity.ID)
	}
	if identity.Email != "test@example.com" {
		t.Errorf("expected email=test@example.com, got %v", identity.Email)
	}
	if identity.Extra["plan"] != "basic" {
		t.Errorf("expected plan=basic, got %v", identity.Extra["plan"])
	}
}

func TestUserIdentityEmptyExtra(t *testing.T) {
	identity := UserIdentity{
		ID:    "user-789",
		Email: "test@example.com",
	}

	data, err := json.Marshal(identity)
	if err != nil {
		t.Fatalf("failed to marshal: %v", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}

	if _, ok := result["plan"]; ok {
		t.Error("expected plan field to be absent")
	}
}

func TestSessionMetadataJSON(t *testing.T) {
	metadata := SessionMetadata{
		URL:        "https://example.com/page",
		Referrer:   "https://google.com",
		PageTitle:  "Test Page",
		UserAgent:  "Mozilla/5.0",
		DeviceType: "desktop",
		Browser:    "Chrome",
		OS:         "Windows",
	}

	data, err := json.Marshal(metadata)
	if err != nil {
		t.Fatalf("failed to marshal: %v", err)
	}

	var result SessionMetadata
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}

	if result.URL != "https://example.com/page" {
		t.Errorf("expected url=https://example.com/page, got %v", result.URL)
	}
	if result.PageTitle != "Test Page" {
		t.Errorf("expected pageTitle=Test Page, got %v", result.PageTitle)
	}
}

func TestMessageStatusConstants(t *testing.T) {
	tests := []struct {
		status   MessageStatus
		expected string
	}{
		{MessageStatusSending, "sending"},
		{MessageStatusSent, "sent"},
		{MessageStatusDelivered, "delivered"},
		{MessageStatusRead, "read"},
	}

	for _, tt := range tests {
		if string(tt.status) != tt.expected {
			t.Errorf("expected %s, got %s", tt.expected, tt.status)
		}
	}
}

func TestSenderConstants(t *testing.T) {
	tests := []struct {
		sender   Sender
		expected string
	}{
		{SenderVisitor, "visitor"},
		{SenderOperator, "operator"},
		{SenderAI, "ai"},
	}

	for _, tt := range tests {
		if string(tt.sender) != tt.expected {
			t.Errorf("expected %s, got %s", tt.expected, tt.sender)
		}
	}
}

func TestVersionStatusConstants(t *testing.T) {
	tests := []struct {
		status   VersionStatus
		expected string
	}{
		{VersionStatusOK, "ok"},
		{VersionStatusOutdated, "outdated"},
		{VersionStatusDeprecated, "deprecated"},
		{VersionStatusUnsupported, "unsupported"},
	}

	for _, tt := range tests {
		if string(tt.status) != tt.expected {
			t.Errorf("expected %s, got %s", tt.expected, tt.status)
		}
	}
}

func TestMessageJSON(t *testing.T) {
	now := time.Now()
	delivered := now.Add(-time.Minute)

	msg := Message{
		ID:          "msg-123",
		SessionID:   "sess-456",
		Content:     "Hello, world!",
		Sender:      SenderVisitor,
		Timestamp:   now,
		Status:      MessageStatusDelivered,
		DeliveredAt: &delivered,
	}

	data, err := json.Marshal(msg)
	if err != nil {
		t.Fatalf("failed to marshal: %v", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}

	if result["sessionId"] != "sess-456" {
		t.Errorf("expected sessionId=sess-456, got %v", result["sessionId"])
	}
	if result["sender"] != "visitor" {
		t.Errorf("expected sender=visitor, got %v", result["sender"])
	}
}

func TestSessionJSON(t *testing.T) {
	now := time.Now()

	session := Session{
		ID:             "sess-123",
		VisitorID:      "visitor-456",
		CreatedAt:      now,
		LastActivity:   now,
		OperatorOnline: true,
		AIActive:       false,
	}

	data, err := json.Marshal(session)
	if err != nil {
		t.Fatalf("failed to marshal: %v", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}

	if result["visitorId"] != "visitor-456" {
		t.Errorf("expected visitorId=visitor-456, got %v", result["visitorId"])
	}
	if result["operatorOnline"] != true {
		t.Errorf("expected operatorOnline=true, got %v", result["operatorOnline"])
	}
}

func TestTrackedElementJSON(t *testing.T) {
	elem := TrackedElement{
		Selector:      ".pricing-btn",
		Event:         "click",
		Name:          "clicked_pricing",
		WidgetMessage: "Interested in our pricing?",
		Data: map[string]interface{}{
			"plan": "pro",
		},
	}

	data, err := json.Marshal(elem)
	if err != nil {
		t.Fatalf("failed to marshal: %v", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}

	if result["selector"] != ".pricing-btn" {
		t.Errorf("expected selector=.pricing-btn, got %v", result["selector"])
	}
	if result["widgetMessage"] != "Interested in our pricing?" {
		t.Errorf("expected widgetMessage, got %v", result["widgetMessage"])
	}
}

func TestCustomEventJSON(t *testing.T) {
	now := time.Now()

	event := CustomEvent{
		Name:      "test_event",
		Data:      map[string]interface{}{"key": "value"},
		Timestamp: now,
		SessionID: "sess-123",
	}

	data, err := json.Marshal(event)
	if err != nil {
		t.Fatalf("failed to marshal: %v", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}

	if result["name"] != "test_event" {
		t.Errorf("expected name=test_event, got %v", result["name"])
	}
	if result["sessionId"] != "sess-123" {
		t.Errorf("expected sessionId=sess-123, got %v", result["sessionId"])
	}
}

func TestValidateContent(t *testing.T) {
	// Valid content
	err := ValidateContent("Hello, world!")
	if err != nil {
		t.Errorf("expected no error for valid content, got %v", err)
	}

	// Content too long
	longContent := make([]byte, MaxMessageContentLength+1)
	for i := range longContent {
		longContent[i] = 'a'
	}
	err = ValidateContent(string(longContent))
	if err != ErrContentTooLong {
		t.Errorf("expected ErrContentTooLong, got %v", err)
	}
}

func TestConnectRequestJSON(t *testing.T) {
	req := ConnectRequest{
		VisitorID: "visitor-123",
		SessionID: "session-456",
		Metadata: &SessionMetadata{
			URL: "https://example.com",
		},
	}

	data, err := json.Marshal(req)
	if err != nil {
		t.Fatalf("failed to marshal: %v", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}

	if result["visitorId"] != "visitor-123" {
		t.Errorf("expected visitorId=visitor-123, got %v", result["visitorId"])
	}
}

func TestConnectResponseJSON(t *testing.T) {
	resp := ConnectResponse{
		SessionID:      "sess-123",
		VisitorID:      "visitor-456",
		OperatorOnline: true,
		WelcomeMessage: "Hello!",
		Messages:       []Message{},
	}

	data, err := json.Marshal(resp)
	if err != nil {
		t.Fatalf("failed to marshal: %v", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}

	if result["welcomeMessage"] != "Hello!" {
		t.Errorf("expected welcomeMessage=Hello!, got %v", result["welcomeMessage"])
	}
}
